File: dotnet/SpiderEngine/Parser/Content/RegexContentParser.cs
namespace SpiderEngine.Parser.Content;

using System.Collections.Generic;
using System.Text.RegularExpressions;
using SpiderEngine.Abstract;
using SpiderEngine.Exception;
using SpiderEngine.Model;

internal class RegexContentParser : IContentParser
{
    public IDictionary<string, string> Context { get; set; } = default!;

    public void Parse(string content, string patten, IEnumerable<SpiderKeyPathPair> value)
    {
        Regex.Match(content, patten);
        var matches = Regex.Matches(content, patten, RegexOptions.Singleline).ToList();

        try
        {
            var dict = matches
                .Select(
                    (match, index) =>
                    {
                        if (!match.Success)
                        {
                            throw new ParseException("Content does not match the patten.");
                        }
                        return (index, match.Groups.Values.First());
                    }
                )
                .ToDictionary();
            foreach (var pair in value)
            {
                Context[pair.Key] = dict[int.Parse(pair.Path)].Value;
            }
        }
        catch (AggregateException ex)
        {
            throw new ParseException("Regex parse failed", [.. ex.InnerExceptions]);
        }
        catch (System.Exception ex)
        {
            throw new ParseException("Content does not match the patten.", ex);
        }
    }
}

File: android/src/main/kotlin/com/itoken/team/wutspider/impl/parser/content/RegexContentParser.kt
package com.itoken.team.wutspider.impl.parser.content

import com.itoken.team.wutspider.SpiderException
import com.itoken.team.wutspider.interfaces.IContentParser
import com.itoken.team.wutspider.model.SpiderKeyPathPair

class RegexContentParser : IContentParser {

    override var context: MutableMap<String, String>? = null

    override fun parse(content: String, pattern: String?, value: List<SpiderKeyPathPair>) {
        val regex = runCatching { Regex(pattern!!, RegexOption.DOT_MATCHES_ALL) }.getOrElse {
            throw SpiderException("Invalid pattern '$pattern'", it)
        }
        val result = regex.find(content) ?: throw SpiderException("Can not find pattern $pattern in content")
        value.forEach {
            val index = it.path.toIntOrNull() ?: throw SpiderException("Invalid group index '$it' for regex")
            context?.put(it.key, result.groupValues[index])
        }

    }


}